src/main/java/life/qbic/subscriptions/SubscriptionController.java

```diff
@@ -0,0 +1,146 @@
+import life.qbic.subscriptions.subscriptions.SubscriptionRepository;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Subscription", description = "Subscription controller API")
+@RestController
+@RequestMapping("/subscription")
+public class SubscriptionController {
+
+  private static final Logger log = LogManager.getLogger(SubscriptionController.class);
+
+  RequestEncrypter requestEncrypter;
+
+  RequestDecrypter requestDecrypter;
+
+  SubscriptionRepository subscriptionRepository;
+
+  @Autowired
+  SubscriptionController(
+      SubscriptionRepository subscriptionRepository,
+      RequestDecrypter requestDecrypter,
+      RequestEncrypter requestEncrypter
+      ) {
+    this.subscriptionRepository = subscriptionRepository;
+    this.requestDecrypter = requestDecrypter;
+    this.requestEncrypter = requestEncrypter;
+  }
+
+  @RequestMapping(value = "/cancel", method = RequestMethod.GET)
+  public ResponseEntity<String> getCancellationRequestHash(
+      @RequestBody CancellationRequest cancellationRequest) {
+    validateRequest(cancellationRequest);
+    var cancellationRequestHash = requestEncrypter.encryptCancellationRequest(cancellationRequest);
+    return new ResponseEntity<>(cancellationRequestHash, HttpStatus.OK);
+  }
+
+  @RequestMapping(value = "/cancel/{hash}", method = RequestMethod.POST)
+  public ResponseEntity<CancellationRequest> cancelSubscription(
+      @PathVariable(value = "hash") String requestHash) {
+    var cancellationRequest = requestDecrypter.decryptCancellationRequest(requestHash);
+    removeSubscription(cancellationRequest);
+    return new ResponseEntity<>(cancellationRequest, HttpStatus.ACCEPTED);
+  }
+
+  private void removeSubscription(CancellationRequest request) {
+    try {
+      subscriptionRepository.cancelSubscription(request);
+    } catch (Exception e) {
+      log.error(e);
+      throw new CancellationFailure("Unexpected failure.");
+    }
+  }
+
+  private void validateRequest(CancellationRequest cancellationRequest) {
+    var project = cancellationRequest.project();
+    var userId = cancellationRequest.userId();
+    if (project == null || project.isBlank()) {
+      throw new MissingPropertyException("project");
+    }
+    if (userId == null || userId.isBlank()) {
+      throw new MissingPropertyException("userId");
+    }
+  }
+
+  @ExceptionHandler({EncryptionException.class})
```

Comment:
Does that mean that this method is called when this exception is thrown within this class?

Reply:
exactly! how awesome is that, isn't it?